### ISSUE TITLE
[Bug]: Using the column table caused inconsistencies between the btree index and the column table data.

### ISSUE THREAD

**Issue by ZhiXingHeYiApple** [hydradatabase/hydra#216]
### What's wrong?

Since in my scenario, there will only be a single write scenario, so I used column partitioned table, below is my table structure：
```sql
CREATE TABLE strategic_developer_api_access_data (
	id bigserial NOT NULL,
	data_source varchar NULL,
	stat_date date NULL,
	stat_time int8 NULL,
	developer_id varchar(100) NULL,
	api_type varchar(150) NULL,
	api_path varchar(300) NULL,
	api_name varchar(100) NULL,
	res_code varchar(10) NULL,
	count int4 NULL,
	max_cost float4 NULL,
	min_cost float4 NULL,
	avg_cost float4 NULL,
	cost1 int4 NULL DEFAULT 0,
	cost12 int4 NULL DEFAULT 0,
	cost23 int4 NULL DEFAULT 0,
	cost34 int4 NULL DEFAULT 0,
	cost45 int4 NULL DEFAULT 0,
	cost5x int4 NULL DEFAULT 0,
	costx int4 NULL DEFAULT 0,
	avg_tps int4 NULL DEFAULT 0,
	max_tps int4 NULL DEFAULT 0,
	create_time timestamp(3) NULL DEFAULT now(),
	update_time timestamp(3) NULL DEFAULT now(),
	CONSTRAINT strategic_developer_api_access_data_pk PRIMARY KEY (id，stat_date, developer_id)
) PARTITION BY RANGE (stat_date);


-- columnar partition
CREATE TABLE strategic_developer_api_access_data_202312 PARTITION OF strategic_developer_api_access_data
FOR VALUES FROM ('2023-12-01') TO ('2024-01-01') partition by hash (developer_id);

CREATE TABLE strategic_developer_api_access_data_202312_hash_p1
PARTITION OF strategic_developer_api_access_data_202312
FOR VALUES WITH (modulus 5, remainder 0) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202312_hash_p2
PARTITION OF strategic_developer_api_access_data_202312
FOR VALUES WITH (modulus 5, remainder 1) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202312_hash_p3
PARTITION OF strategic_developer_api_access_data_202312
FOR VALUES WITH (modulus 5, remainder 2) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202312_hash_p4
PARTITION OF strategic_developer_api_access_data_202312
FOR VALUES WITH (modulus 5, remainder 3) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202312_hash_p5
PARTITION OF strategic_developer_api_access_data_202312
FOR VALUES WITH (modulus 5, remainder 4) USING columnar;

-- next month sub partition table
CREATE TABLE strategic_developer_api_access_data_202401 PARTITION OF strategic_developer_api_access_data
FOR VALUES FROM ('2024-01-01') TO ('2024-02-01') partition by hash (developer_id);

CREATE TABLE strategic_developer_api_access_data_202401_hash_p1
PARTITION OF strategic_developer_api_access_data_202401
FOR VALUES WITH (modulus 5, remainder 0) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202401_hash_p2
PARTITION OF strategic_developer_api_access_data_202401
FOR VALUES WITH (modulus 5, remainder 1) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202401_hash_p3
PARTITION OF strategic_developer_api_access_data_202401
FOR VALUES WITH (modulus 5, remainder 2) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202401_hash_p4
PARTITION OF strategic_developer_api_access_data_202401
FOR VALUES WITH (modulus 5, remainder 3) USING columnar;

CREATE TABLE strategic_developer_api_access_data_202401_hash_p5
PARTITION OF strategic_developer_api_access_data_202401
FOR VALUES WITH (modulus 5, remainder 4) USING columnar;

```

In the process of writing data to the table `strategic_developer_api_access_data`, there will be operations of first delete by primary keys（`id`、`stat_date`、`developer_id`） and then insert. After the program has been running for a while, I encountered an inconsistency between the `btree` index and the column table data.

```sql
set enable_indexscan to off;
select * from strategic_developer_api_access_data where id=5507 and stat_date ='2023-12-20'::date and developer_id ='1575882239013393365';

QUERY PLAN                                                                                                                                                                                                                                                     |
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Custom Scan (ColumnarScan) on strategic_developer_api_access_data_202312_hash_p3 strategic_developer_api_access_data  (cost=0.00..31.93 rows=315 width=221) (actual time=0.775..1.088 rows=1 loops=1)                                                          |
  Filter: ((developer_id)::text = '1575882239013393365'::text)                                                                                                                                                                                                 |
  Columnar Projected Columns: id, data_source, stat_date, stat_time, developer_id, api_type, api_path, api_name, res_code, count, max_cost, min_cost, avg_cost, cost1, cost12, cost23, cost34, cost45, cost5x, costx, avg_tps, max_tps, create_time, update_tim|
  Columnar Chunk Group Filters: ((id = 5507) AND (stat_date = '2023-12-20'::date))                                                                                                                                                                             |
  Columnar Chunk Groups Removed by Filter: 3                                                                                                                                                                                                                   |
  Columnar Vectorized Filter: ((id = 5507) AND (stat_date = '2023-12-20'::date))                                                                                                                                                                               |
Planning Time: 0.288 ms                                                                                                                                                                                                                                        |
Execution Time: 1.391 ms                                                                                                                                                                                                                                       |

8 row(s) fetched.
```

If I `set enable_indexscan to on`, the above query will use the btree index, It is impossible to query that row of records. The query analyze plan is below:
```sql
QUERY PLAN                                                                                                                                                                                                                                      |
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Index Scan using strategic_developer_api_access_data_202312_hash_p3_pkey on strategic_developer_api_access_data_202312_hash_p3 strategic_developer_api_access_data  (cost=0.27..8.73 rows=1 width=221) (actual time=0.014..0.014 rows=0 loops=1)|
  Index Cond: ((id = 5507) AND (stat_date = '2023-12-20'::date) AND ((developer_id)::text = '1575882239013393365'::text))                                                                                                                       |
Planning Time: 0.297 ms                                                                                                                                                                                                                         |
Execution Time: 0.027 ms                                                                                                                                                                                                                        |

4 row(s) fetched.
```

**Comment by JerrySievert** [comment]
do you have `amcheck` installed?  (`create extension amcheck;`)

if so, can you please run `SELECT bt_index_check('strategic_developer_api_access_data_202312_hash_p3_pkey');` and check your logs for any errors or warnings?

that should check to see if the index has become corrupt.

**Comment by wuputah** [comment]
Not sure what version you are using but is it possible this was fixed in #182, but the data was inserted before that fix? That may also impact the results. I would think rebuilding the index would fix the problem.

**Comment by wuputah** [comment]
Closing as stale.
